src/lib/replication/repl_dev/raft_repl_dev.cpp

```diff
@@ -1103,15 +1104,31 @@ void RaftReplDev::cp_flush(CP* cp) {
 
     std::unique_lock lg{m_sb_mtx};
     m_rd_sb->compact_lsn = clsn;
-    m_rd_sb->durable_commit_lsn = lsn;
+    // dc_lsn is also flushed in flush_durable_commit_lsn()
+    // we need to take a max to avoid rolling back.
+    m_rd_sb->durable_commit_lsn = std::max(lsn, m_rd_sb->durable_commit_lsn);
```

Comment:
here, lsn is equal to m_commit_upto_lsn.load(), and m_rd_sb->durable_commit_lsn should always be equal or less than lsn, right?
in which case, lsn is bigger than  m_rd_sb->durable_commit_lsn?

Reply:
the lsn is from L1096, which was captured from the state_machine during cp_switchover(say LSN=100)
There can be some commits happen in the state machine afterward,  when we reaching here, the latest LSN from state machine can be 140,  but the dc_lsn, persisted every 500ms by reaper_thread, can be 120.

Reply:
sorry , I have give a wrong question in my above comment. 
I was planning to ask in which case lsn is smaller than m_rd_sb->durable_commit_lsn?

Reply:
if CP happens right after dc_lsn persist. DC_LSN persist to 100,  we made 2 more commits then do a CP:)

Reply:
>if CP happens right after dc_lsn persist. DC_LSN persist to 100, we made 2 more commits then do a CP:)

this also means m_rd_sb->durable_commit_lsn should always be equal or less than lsn, right?
I mean we can do `m_rd_sb->durable_commit_lsn = lsn` , this is enough.

Reply:
Let me try to explain this one more time.

We capture LSN from state-machine during cp_switchover(say LSN=100 at that time), saved into cp ctx.
During CP flush  here , the we use lsn:100 to update cp_lsn and maybe dc_lsn.


However, there is a time window, between cp_switchover and cp_flush,  the state machine still moving ahead, it can be 120 when we reaching cp_flush. And we have reaper_thread which persist dc_lsn , it can happen at the moment to persist 120.  So at the moment,  `m_rd_sb->durable_commit_lsn == 120`

Now the situation is:
state machine has committed up to 120
index and other component (cp) persist up to 100.

We should make cp_lsn = 100 and keep dc_lsn = 120.

Reply:
ok , make sense